config/constructor_types.go

```diff
@@ -28,43 +29,47 @@ var (
 	connGaterType = reflect.TypeOf((*connmgr.ConnectionGater)(nil)).Elem()
 	upgraderType  = reflect.TypeOf((*transport.Upgrader)(nil)).Elem()
 	rcmgrType     = reflect.TypeOf((*network.ResourceManager)(nil)).Elem()
+	resolverType  = reflect.TypeOf((*madns.Resolver)(nil)).Elem()
```

Comment:
Would be nice to use an interface instead of a concrete type here. `madns.BasicResolver`?

Could it be that `madns.Resolver` and `madns.BasicResolver` should switch their names?

Reply:
So, `madns.Resolver` is basically a light-weight wrapper over a set of `BasicResolver`s that implements a multiaddr handing [`Resolve`](https://github.com/multiformats/go-multiaddr-dns/blob/0bad63ddb983cf7824f45f1490f81f3b3b66fe62/resolve.go#L108) method. The pluggable component is really `madns.BasicResolver`, where the user should then construct an `madns.Resolver` from one or more `madns.BasicResolver`s.